### PR TITLE
Initialize PoS Chain At Startup To Continue Existing Chain

### PR DESCRIPTION
**Pull request by tholonious**

### CODE REVIEW

**Comment by tholonious** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1031?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `feature/proof-of-stake`
  * **PR #1028** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1028?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #1029** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1029?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1030** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1030?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #1031** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1031?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/1031?utm_source=stack-comment).

File: lib/blockchain.go [hunk 1]
```diff
@@ -700,6 +700,47 @@ func (bc *Blockchain) _initChain() error {
...
+		return !node.IsCommitted()
+	})
+
+	// If there are no uncommitted blocks, we're done.
+	if len(safeBlockNodes) == 0 {
+		return nil
+	}
+
+	// Find the safe block with the highest view. That block is the uncommitted tip.
+	uncommittedTipBlocNode := safeBlockNodes[0]
```
Comment (lazynina) [inline lib/blockchain.go:723]:
```suggestion
	uncommittedTipBlockNode := safeBlockNodes[0]
```

File: lib/blockchain.go [hunk 2]
```diff
@@ -700,6 +700,47 @@ func (bc *Blockchain) _initChain() error {
...
+	for _, blockNode := range safeBlockNodes {
+		if blockNode.Header.ProposedInView > uncommittedTipBlocNode.Header.ProposedInView {
+			uncommittedTipBlocNode = blockNode
+		}
+	}
+
+	// Fetch the lineage of blocks from the committed tip through the uncommitted tip.
+	lineageFromCommittedTip, err := bc.getLineageFromCommittedTip(uncommittedTipBlocNode.Header)
+	if err != nil {
+		return errors.Wrapf(err, "NewBlockchain: ")
```
Comment (lazynina) [inline lib/blockchain.go:733]:
```suggestion
		return errors.Wrapf(err, "_applyUncommittedBlocksToBestChain: ")
```

File: lib/blockchain.go [hunk 3]
```diff
@@ -700,6 +700,47 @@ func (bc *Blockchain) _initChain() error {
...
+	}
+
+	// Add the uncommitted blocks to the in-memory data structures.
+	if _, _, _, err := bc.tryApplyNewTip(uncommittedTipBlocNode, 0, lineageFromCommittedTip); err != nil {
+		return errors.Wrapf(err, "NewBlockchain: ")
```
Comment (lazynina) [inline lib/blockchain.go:738]:
```suggestion
		return errors.Wrapf(err, "_applyUncommittedBlocksToBestChain: ")
```

**Review by lazynina — APPROVED** [review]
Just some nits on naming and error messages, but otherwise, LGTM.

**Comment by tholonious** [comment]
### Merge activity

* **Feb 13, 7:09 PM EST**: @tholonious started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/1031).

**Pull request merged**